### PR TITLE
Add pluggable datasource support

### PR DESCRIPTION
**Pull request by zetcco**
$subject

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,141 @@
...
+## Features
+-  **Perform Dry Runs**: Verify a given Subgraph for Supergraph composition before publishing it.
+-   **Publish Subgraph**: Publish a Subgraph to the Schema Registry for Supergraph composition.
+-   **Prevent Breaking Changes**: Ensure Subgraphs don't cause breaking changes to the Supergraph.
+-   **Automatic Versioning**: Automatically versions the Supergraph based on the changes that occurred to it in the format of `[Breaking].[Dangerous].[Safe]`. (Similar to Semantic Versioning)
+-   **Get Supergraph Schema/API Schema**: Access Supergraph and its API Schema.
+-   **Get Supergraph Differences**: Compare two Supergraphs and generate a diff between the two.
+- **Categorize Diff Type** - Automatically categorize the changes occured to a Schema into **Breaking**, **Dangerous** and **Safe**.
+- **Data-Agnostic**: Allows any datasource integration. (MongoDB, File-based storage and In-memory storage are supported by default)
+-   **Get Registered Subgraph by Name**: Retrieve Subgraph from Registry.
```
Comment (ThisaruGuruge) [inline README.md:14]: Better to use a proper markdown formatter and/or linter.

```suggestion
- **Perform Dry Runs**: Verify a given Subgraph for Supergraph composition before publishing it.
- **Publish Subgraph**: Publish a Subgraph to the Schema Registry for Supergraph composition.
- **Prevent Breaking Changes**: Ensure Subgraphs don't cause breaking changes to the Supergraph.
- **Automatic Versioning**: Automatically versions the Supergraph based on the changes that occurred to it in the format of `[Breaking].[Dangerous].[Safe]`. (Similar to Semantic Versioning)
- **Get Supergraph Schema/API Schema**: Access Supergraph and its API Schema.
- **Get Supergraph Differences**: Compare two Supergraphs and generate a diff between the two.
- **Categorize Diff Type** - Automatically categorize the changes occured to a Schema into **Breaking**, **Dangerous** and **Safe**.
- **Data-Agnostic**: Allows any datasource integration. (MongoDB, File-based storage and In-memory storage are supported by default)
- **Get Registered Subgraph by Name**: Retrieve Subgraph from Registry.
```

File: ballerina/service.bal [hunk 2]
```diff
@@ -15,54 +15,62 @@
...
+    return schemaRegistryService;
+}
+
+public function main() returns error? {
+    graphql:Listener graphqlListener = check new (port);
+    graphql:Service registryService = getSchemaRegistryService(check new MongodbDatasource());
+    check graphqlListener.attach(registryService);
+    check graphqlListener.'start();
+    runtime:registerListener(graphqlListener);
 }
```
Comment (ThisaruGuruge) [inline ballerina/service.bal:76]: We should provide a public function other than the `main` function. Then we can publish this to the Ballerina central, and the users can run the server by importing this package.

File: ballerina/service.bal [hunk 3]
```diff
@@ -15,54 +15,66 @@
...
+    };
+    return schemaRegistryService;
+}
+
+public function startRegistry(datasource:Datasource datasource, int port) returns error? {
+    graphql:Listener graphqlListener = check new (port);
+    graphql:Service registryService = getSchemaRegistryService(datasource);
+    check graphqlListener.attach(registryService);
+    check graphqlListener.'start();
+    runtime:registerListener(graphqlListener);
```
Comment (ThisaruGuruge) [inline ballerina/service.bal:75]: Shouldn't we register the listener first?

```suggestion
    runtime:registerListener(graphqlListener);
    check graphqlListener.attach(registryService);
    check graphqlListener.'start();
```

Comment (zetcco) [reply ballerina/service.bal:75]: I had the same problem, but I decided to follow these ([dynamic http listener](https://ballerina.io/learn/by-example/dynamic-listener/), [dynamic graphql listener test](https://github.com/ballerina-platform/module-ballerina-graphql/blob/63450ac938c84fb98338a6a584b87f030c01bce0/compiler-plugin-tests/src/test/resources/ballerina_sources/validator_tests/23_dynamically_attaching_service/service.bal#L39)) references. What should we do?

Comment (ThisaruGuruge) [reply ballerina/service.bal:75]: Hmm, right. That makes sense. Let's keep it as it is.

**Review by ThisaruGuruge — APPROVED** [review]

**Pull request merged**